internal/ledgerutil/verify/verify.go

```diff
@@ -0,0 +1,224 @@
+			result, err := iterator.Next()
+			if err != nil {
+				return false, err
+			}
+			if result == nil {
+				break
+			}
+			block, ok := result.(*common.Block)
+			if !ok {
+				return false, errors.Errorf("Cannot decode the block %d", i)
+			}
+
+			// Perform checks for this block
+			checkResult, err := checkBlock(block, previousHash)
+			if err != nil {
+				return false, err
+			}
+
+			// If any error is found, flag as so, and vice versa
+			if !checkResult.Valid {
+				noLedgerError = false
+				blockResult.ErroneousBlocks += 1
+
+				// Record the error details in the JSON
+				err = blockResult.Writer.AddEntry(checkResult)
+				if err != nil {
+					return false, err
+				}
+			} else {
+				blockResult.ValidBlocks += 1
+			}
+
+			// Remember the header hash to compare it with the "previous hash" in the next block
+			previousHash = protoutil.BlockHeaderHash(block.Header)
+		}
+
+		err = blockResult.Writer.CloseList()
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// Returns true only if no error is found in all the channels
+	return noLedgerError, nil
+}
+
+// Get a default block store provider to access the peer's block store
+func getBlockStoreProvider(fsPath string) (*blkstorage.BlockStoreProvider, error) {
+	// Format path to block store
+	blockStorePath := kvledger.BlockStorePath(filepath.Join(fsPath, ledgersDataDirName))
+	isEmpty, err := fileutil.DirEmpty(blockStorePath)
+	if err != nil {
+		return nil, err
+	}
+	if isEmpty {
+		return nil, errors.Errorf("provided path %s is empty. Aborting verify", fsPath)
+	}
+	// Default fields for block store provider
+	conf := blkstorage.NewConf(blockStorePath, 0)
+	indexConfig := &blkstorage.IndexConfig{
+		AttrsToIndex: []blkstorage.IndexableAttr{
+			blkstorage.IndexableAttrBlockNum,
+			blkstorage.IndexableAttrBlockHash,
+			blkstorage.IndexableAttrTxID,
+			blkstorage.IndexableAttrBlockNumTranNum,
+		},
+	}
+	metricsProvider := &disabled.Provider{}
+	// Create new block store provider
+	blockStoreProvider, err := blkstorage.NewProvider(conf, indexConfig, metricsProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	return blockStoreProvider, nil
+}
+
+// blockResultOutput - stores the number of valid blocks and those of erroneous blocks
+// and the JSON writer object
+type blockResultOutput struct {
+	ValidBlocks     uint64
+	ErroneousBlocks uint64
+	Writer          *jsonrw.JSONFileWriter
+}
+
+func newBlockResultOutput(directory string) (*blockResultOutput, error) {
+	writer, err := jsonrw.NewJSONFileWriter(filepath.Join(directory, blockResultJsonFilename))
+	if err != nil {
+		return nil, err
+	}
+
+	return &blockResultOutput{
+		ValidBlocks:     0,
+		ErroneousBlocks: 0,
+		Writer:          writer,
+	}, nil
+}
+
+// blockCheckResult - stores the result of the checks for a block
+type blockCheckResult struct {
+	Number uint64   `json:"number"`
+	Valid  bool     `json:"valid"`
+	Errors []string `json:"errors,omitempty"`
+}
+
+// Perform checks on a block
```

Comment:
```suggestion
// checkBlock checks a block's hash
```

Not as big a deal for unexported functions, but still the function comment should start with the function name.